types/src/transaction/authenticator.rs

```diff
@@ -76,84 +84,184 @@ impl TransactionAuthenticator {
         }
     }
 
-    /// Return bytes for (self.public_key | self.scheme_id). To create an account authentication
-    /// key, sha3 these bytes
-    // TODO: move AuthenticationKey to this file and make this private
-    pub fn compute_authentication_key_preimage(
-        mut public_key_bytes: Vec<u8>,
-        scheme_id: u8,
-    ) -> AuthenticationKeyPreimage {
-        public_key_bytes.push(scheme_id);
-        AuthenticationKeyPreimage(public_key_bytes)
+    /// Return an authentication key preimage derived from `self`'s public key and scheme id
+    pub fn authentication_key_preimage(&self) -> AuthenticationKeyPreimage {
+        AuthenticationKeyPreimage::new(self.public_key_bytes(), self.scheme())
     }
 
-    // TODO: move AuthenticationKey to this file to avoid try_from
-    fn preimage_to_authentication_key(preimage: &AuthenticationKeyPreimage) -> AuthenticationKey {
-        AuthenticationKey::try_from(HashValue::from_sha3_256(&preimage.0).to_vec()).unwrap()
+    /// Return an authentication key derived from `self`'s public key and scheme id
+    pub fn authentication_key(&self) -> AuthenticationKey {
+        AuthenticationKey::from_preimage(&self.authentication_key_preimage())
     }
+}
 
-    pub fn authentication_key_preimage(&self) -> AuthenticationKeyPreimage {
-        Self::compute_authentication_key_preimage(
-            self.public_key_bytes().to_vec(),
-            self.scheme_id(),
-        )
+/// A struct that represents an account authentication key. An account's address is the last 16
+/// bytes of authentication key used to create it
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy, CryptoHasher)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+pub struct AuthenticationKey([u8; AUTHENTICATION_KEY_LENGTH]);
+
+impl AuthenticationKey {
+    /// Create an authentication key from `bytes`
+    pub const fn new(bytes: [u8; AUTHENTICATION_KEY_LENGTH]) -> Self {
+        Self(bytes)
+    }
+
+    /// Create an authentication key from a preimage by taking its sha3 hash
+    pub fn from_preimage(preimage: &AuthenticationKeyPreimage) -> AuthenticationKey {
+        AuthenticationKey::new(*HashValue::from_sha3_256(&preimage.0).as_ref())
+    }
+
+    /// Create an authentication key from an Ed25519 public key
+    pub fn ed25519(public_key: &Ed25519PublicKey) -> AuthenticationKey {
+        Self::from_preimage(&AuthenticationKeyPreimage::ed25519(public_key))
+    }
+
+    /// Create an authentication key from a MultiEd25519 public key
+    pub fn multi_ed25519(public_key: &MultiEd25519PublicKey) -> Self {
+        Self::from_preimage(&AuthenticationKeyPreimage::multi_ed25519(public_key))
+    }
+
+    /// Return an address derived from the last ADDRESS_LENGTH bytes of this authentication key
+    pub fn derived_address(&self) -> AccountAddress {
+        // keep only last 16 bytes
+        let mut array = [0u8; ADDRESS_LENGTH];
```

Comment:
Samething ADDRESS_LENGTH: AccountAddress::LENGTH, but @sunshowers might be on top of that one...